lib/backend/dynamo/shards.go

```diff
@@ -83,26 +83,70 @@ func (b *Backend) pollStreams(externalCtx context.Context) error {
 	set := make(map[string]struct{})
 	eventsC := make(chan shardEvent)
 
-	refreshShards := func() error {
+	shouldStartPoll := func(shard *dynamodbstreams.Shard) bool {
+		sid := aws.StringValue(shard.ShardId)
+		if _, ok := set[sid]; ok {
+			// already being polled
+			return false
+		}
+		if _, ok := set[aws.StringValue(shard.ParentShardId)]; ok {
+			b.Warnf("Skipping child shard: %s, still polling parent %s", sid, aws.StringValue(shard.ParentShardId))
+			// still processing parent
+			return false
+		}
+		return true
+	}
+
+	refreshShards := func(init bool) (bool, error) {
 		shards, err := b.collectActiveShards(ctx, streamArn)
 		if err != nil {
-			return trace.Wrap(err)
+			return false, trace.Wrap(err)
+		}
+
+		var initC chan error
+		if init {
+			// first call to  refreshShards requires us to block on shard iterator
+			// registration.
+			initC = make(chan error, len(shards))
 		}
+
+		started := 0
 		for i := range shards {
+			if !shouldStartPoll(shards[i]) {
+				continue
+			}
 			shardID := aws.StringValue(shards[i].ShardId)
-			if _, ok := set[shardID]; !ok {
-				b.Debugf("Adding active shard %v.", shardID)
-				set[shardID] = struct{}{}
-				go b.asyncPollShard(ctx, streamArn, shards[i], eventsC)
+			b.Debugf("Adding active shard %v.", shardID)
+			set[shardID] = struct{}{}
+			go b.asyncPollShard(ctx, streamArn, shards[i], eventsC, initC)
+			started++
+		}
+
+		if init {
+			// block on shard iterator registration.
+			for i := 0; i < started; i++ {
+				select {
+				case err = <-initC:
+					if err != nil {
+						return false, trace.Wrap(err)
+					}
+				case <-ctx.Done():
+					return true, nil
```

Comment:
The return type of `refreshShards` should perhaps be just `error`, with `ctx.Err()` being returned here instead of `true, nil`. As it stands, when the `closing` condition is encountered below this cancellation info is simply discarded.

Reply:
Good point.  The `closing` bool used to be used to suppress spurious logs during an earlier iteration of this code.  Since then I changed to checking `isClosed()` directly before emitting the error log, so it isn't needed anymore.